src/analysis/calculus/deriv.lean

```diff
@@ -1176,3 +1206,68 @@ begin
 end
 
 end
+
+section tangent_cone
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{F : Type*} [normed_group F] [normed_space 𝕜 F]
+{f : E → F} {s : set E} {f' : E →L[𝕜] F}
+
+/-- The image of a tangent cone under the differential of a map is included in the tangent cone to
+the image. -/
+lemma has_fderiv_within_at.image_tangent_cone_subset {x : E} (h : has_fderiv_within_at f f' s x) :
+  f' '' (tangent_cone_at 𝕜 s x) ⊆ tangent_cone_at 𝕜 (f '' s) (f x) :=
+begin
+  rw image_subset_iff,
+  rintros v ⟨c, d, dtop, clim, cdlim⟩,
+  refine ⟨c, (λn, f (x + d n) - f x), _, clim, _⟩,
+  show {n : ℕ | f x + (f (x + d n) - f x) ∈ f '' s} ∈ at_top,
+  { apply mem_sets_of_superset dtop,
+    simp [-mem_image, mem_image_of_mem] {contextual := tt} },
+  show tendsto (λ (n : ℕ), c n • (f (x + d n) - f x)) at_top (nhds (f' v)),
+  { apply h.lim dtop clim cdlim, }
+end
+
+/-- If a set has the unique differentiability property at a point x, then the image of this set
+under a map with onto derivative has also the unique differentiability property at the image point.
+-/
+lemma has_fderiv_within_at.unique_diff_within_at {x : E} (h : has_fderiv_within_at f f' s x)
+  (hs : unique_diff_within_at 𝕜 s x) (h' : closure (range f') = univ) :
+  unique_diff_within_at 𝕜 (f '' s) (f x) :=
+begin
+  have A : ∀v ∈ tangent_cone_at 𝕜 s x, f' v ∈ tangent_cone_at 𝕜 (f '' s) (f x),
+  { assume v hv,
+    have := h.image_tangent_cone_subset,
+    rw image_subset_iff at this,
+    exact this hv },
+  have B : ∀v ∈ (submodule.span 𝕜 (tangent_cone_at 𝕜 s x) : set E),
+    f' v ∈ (submodule.span 𝕜 (tangent_cone_at 𝕜 (f '' s) (f x)) : set F),
+  { assume v hv,
+    apply submodule.span_induction hv,
+    { exact λ w hw, submodule.subset_span (A w hw) },
+    { simp },
+    { assume w₁ w₂ hw₁ hw₂,
+      rw continuous_linear_map.map_add,
+      exact submodule.add_mem (submodule.span 𝕜 (tangent_cone_at 𝕜 (f '' s) (f x))) hw₁ hw₂ },
+    { assume a w hw,
+      rw continuous_linear_map.map_smul,
+      exact submodule.smul_mem (submodule.span 𝕜 (tangent_cone_at 𝕜 (f '' s) (f x))) _ hw } },
+  rw [unique_diff_within_at, ← univ_subset_iff],
+  split,
+  show f x ∈ closure (f '' s), from h.continuous_within_at.mem_closure_image hs.2,
+  show univ ⊆ closure ↑(submodule.span 𝕜 (tangent_cone_at 𝕜 (f '' s) (f x))), from calc
```

Comment:
Do you really need to announce the result since it is already explicit thanks to the calc block? Also I feel the `begin ... end` at the end is unaesthetic. I would write:
```lean
  show f x ∈ closure (f '' s), from h.continuous_within_at.mem_closure_image hs.2,
  calc
    univ ⊆ closure (range f') : univ_subset_iff.2 h'
    ... = closure (f' '' univ) : by rw image_univ
    ... = closure (f' '' (closure (submodule.span 𝕜 (tangent_cone_at 𝕜 s x) : set E))) : by rw hs.1
    ... ⊆ closure (closure (f' '' (submodule.span 𝕜 (tangent_cone_at 𝕜 s x) : set E))) :
      closure_mono (image_closure_subset_closure_image f'.cont)
    ... = closure (f' '' (submodule.span 𝕜 (tangent_cone_at 𝕜 s x) : set E)) : closure_closure
    ... ⊆  closure (submodule.span 𝕜 (tangent_cone_at 𝕜 (f '' s) (f x)) : set F) : 
      closure_mono (image_subset_iff.mpr B)
```
Feel free to pick up bits of it if you like some.

Reply:
I kept the initial `show`, as I think it helps in a computation to know where one is heading to, but I removed the final `begin ... block` as you indicated.